File: ios/Namu/Native/Sha256Streamer.swift
import CryptoKit
import Foundation

/// Streaming SHA-256 (DL-010, contract §6.4 step 2). The file is read through
/// one reusable 8 MiB window; hash state is never serialized (DL-004).
enum Sha256Streamer {
  static let bufferSize = 8 * 1024 * 1024

  struct Cancelled: Error {}

  /// - Parameters:
  ///   - progress: called with the total number of bytes hashed so far.
  ///   - isCancelled: polled between buffers; throws `Cancelled` when true.
  /// - Returns: lower-case hex digest.
  static func hashFile(
    at url: URL,
    bufferSize: Int = Sha256Streamer.bufferSize,
    progress: ((Int64) -> Void)? = nil,
    isCancelled: (() -> Bool)? = nil
  ) throws -> String {
    let handle = try FileHandle(forReadingFrom: url)
    defer { try? handle.close() }
    var hasher = SHA256()
    var total: Int64 = 0
    while true {
      if isCancelled?() == true { throw Cancelled() }
      // The pool releases each 8 MiB chunk before the next one is read.
      let count: Int = try autoreleasepool {
        guard let chunk = try handle.read(upToCount: bufferSize), !chunk.isEmpty else { return 0 }
        hasher.update(data: chunk)
        return chunk.count
      }
      if count == 0 { break }
      total += Int64(count)
      progress?(total)
    }
    return Data(hasher.finalize()).namuHex
  }
}

File: android/namu-core/src/main/kotlin/org/namuai/offline/core/hash/Sha256Streamer.kt
package org.namuai.offline.core.hash

import org.namuai.offline.core.util.Hex
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.security.MessageDigest

/**
 * Streaming SHA-256 over the complete file (contract §6.4 step 2). Hash state
 * is never serialized: an interrupted verification starts again from byte 0
 * (DL-004).
 */
object Sha256Streamer {
    const val BUFFER_BYTES = 8 * 1024 * 1024

    /**
     * @param isCancelled polled between buffers; when true the result is null.
     * @param onProgress total bytes hashed so far, called after every buffer.
     * @return lower-case hex digest, or null when cancelled.
     */
    @Throws(IOException::class)
    fun hashFile(
        file: File,
        isCancelled: () -> Boolean = { false },
        onProgress: (Long) -> Unit = {},
        bufferBytes: Int = BUFFER_BYTES,
    ): String? {
        val digest = MessageDigest.getInstance("SHA-256")
        val buffer = ByteArray(bufferBytes)
        var total = 0L
        FileInputStream(file).use { input ->
            while (true) {
                if (isCancelled()) return null
                val n = input.read(buffer)
                if (n < 0) break
                if (n > 0) {
                    digest.update(buffer, 0, n)
                    total += n
                    onProgress(total)
                }
            }
        }
        return Hex.encode(digest.digest())
    }
}
